File: apps/Spark/Spark/SparkKit/Operations/Deposit.swift
//
//  Deposit.swift
//  Spark
//
//  Created by Zhen Lu on 12/18/24.
//  Copyright © 2024 Lightspark Group, Inc. All rights reserved.
//

import Foundation
import secp256k1

func generateDepositAddress(
    client: Spark_SparkServiceAsyncClient,
    identityPublicKey: Data,
    signingPublicKey: Data
) async throws -> Spark_Address {
    var request = Spark_GenerateDepositAddressRequest()
    print("Create Address: \(identityPublicKey.hexString())")
    request.identityPublicKey = identityPublicKey;
    request.signingPublicKey = signingPublicKey;
    let response = try await client.generate_deposit_address(request)
    return response.depositAddress;
}

func createTree(
    client: Spark_SparkServiceAsyncClient,
    onchainTx: Data,
    onchainTxId: String,
    vout: UInt32,
    address: String,
    network: String,
    signingPrivateKey: Data,
    identityPublicKey: Data,
    verifyingPublicKey: Data
) async throws -> Spark_FinalizeNodeSignaturesResponse {
    let signingPublicKey = try secp256k1.Signing.PrivateKey(dataRepresentation: signingPrivateKey).publicKey
    let nodeTxResult = try constructNodeTx(tx: onchainTx, vout: vout, address: address, locktime: 0)
    let refundTxResult = try constructRefundTx(
        tx: nodeTxResult.tx,
        vout: 0,
        pubkey: signingPublicKey.dataRepresentation,
        network: network,
        locktime: 60000
    )

    let keyPackage = KeyPackage(
        secretKey: signingPrivateKey,
        publicKey: signingPublicKey.dataRepresentation,
        verifyingKey: verifyingPublicKey
    )
    let rootNonce = try frostNonce(keyPackage: keyPackage)
    let refundNonce = try frostNonce(keyPackage: keyPackage)

    var request = Spark_StartDepositTreeCreationRequest()
    request.identityPublicKey = identityPublicKey

    request.onChainUtxo.txid = onchainTxId
    request.onChainUtxo.vout = vout

    request.rootTxSigningJob.rawTx = nodeTxResult.tx
    request.rootTxSigningJob.signingPublicKey = signingPublicKey.dataRepresentation
    request.rootTxSigningJob.signingNonceCommitment.hiding = rootNonce.commitment.hiding
    request.rootTxSigningJob.signingNonceCommitment.binding = rootNonce.commitment.binding

    request.refundTxSigningJob.rawTx = refundTxResult.tx
    request.refundTxSigningJob.signingPublicKey = signingPublicKey.dataRepresentation
    request.refundTxSigningJob.signingNonceCommitment.hiding = refundNonce.commitment.hiding
    request.refundTxSigningJob.signingNonceCommitment.binding = refundNonce.commitment.binding
    let response = try await client.start_deposit_tree_creation(request)

    let rootSECommitments = response.rootNodeSignatureShares.nodeTxSigningResult.signingNonceCommitments.mapValues {
        value in
        SigningCommitment(hiding: value.hiding, binding: value.binding)
    }

    let rootUserSig = try signFrost(
        msg: nodeTxResult.sighash,
        keyPackage: keyPackage,
        nonce: rootNonce.nonce,
        selfCommitment: rootNonce.commitment,
        statechainCommitments: rootSECommitments
    )

    let rootSig = try aggregateFrost(
        msg: nodeTxResult.sighash,
        statechainCommitments: rootSECommitments,
        selfCommitment: rootNonce.commitment,
        statechainSignatures: response.rootNodeSignatureShares.nodeTxSigningResult.signatureShares,
        selfSignature: rootUserSig,
        statechainPublicKeys: response.rootNodeSignatureShares.nodeTxSigningResult.publicKeys,
        selfPublicKey: signingPublicKey.dataRepresentation,
        verifyingKey: verifyingPublicKey
    )

    let refundSECommitments = response.rootNodeSignatureShares.refundTxSigningResult.signingNonceCommitments.mapValues {
        value in
        SigningCommitment(hiding: value.hiding, binding: value.binding)
    }

    let refundUserSig = try signFrost(
        msg: refundTxResult.sighash,
        keyPackage: keyPackage,
        nonce: refundNonce.nonce,
        selfCommitment: refundNonce.commitment,
        statechainCommitments: refundSECommitments
    )

    let refundSig = try aggregateFrost(
        msg: refundTxResult.sighash,
        statechainCommitments: refundSECommitments,
        selfCommitment: refundNonce.commitment,
        statechainSignatures: response.rootNodeSignatureShares.refundTxSigningResult.signatureShares,
        selfSignature: refundUserSig,
        statechainPublicKeys: response.rootNodeSignatureShares.refundTxSigningResult.publicKeys,
        selfPublicKey: signingPublicKey.dataRepresentation,
        verifyingKey: verifyingPublicKey
    )

    var finalizeRequest = Spark_FinalizeNodeSignaturesRequest()
    finalizeRequest.intent = .creation
    var nodeSignatures = Spark_NodeSignatures()
    nodeSignatures.nodeID = response.rootNodeSignatureShares.nodeID
    nodeSignatures.nodeTxSignature = rootSig
    nodeSignatures.refundTxSignature = refundSig
    finalizeRequest.nodeSignatures.append(nodeSignatures)

    return try await client.finalize_node_signatures(finalizeRequest)
}

File: sdks/js/packages/spark-sdk/src/services/deposit.ts
import { schnorr, secp256k1 } from "@noble/curves/secp256k1";
import * as btc from "@scure/btc-signer";
import { p2tr, Transaction } from "@scure/btc-signer";
import { equalBytes, sha256 } from "@scure/btc-signer/utils";
import { SignatureIntent } from "../proto/common.js";
import {
  Address,
  FinalizeNodeSignaturesResponse,
  GenerateDepositAddressResponse,
  StartDepositTreeCreationResponse,
} from "../proto/spark.js";
import {
  getP2TRAddressFromPublicKey,
  getSigHashFromTx,
  getTxId,
} from "../utils/bitcoin.js";
import { subtractPublicKeys } from "../utils/keys.js";
import { getNetwork } from "../utils/network.js";
import { proofOfPossessionMessageHashForDepositAddress } from "../utils/proof.js";
import { createWasmSigningCommitment } from "../utils/signing.js";
import { WalletConfigService } from "./config.js";
import { ConnectionManager } from "./connection.js";
type ValidateDepositAddressParams = {
  address: Address;
  userPubkey: Uint8Array;
};

export type GenerateDepositAddressParams = {
  signingPubkey: Uint8Array;
};

export type CreateTreeRootParams = {
  signingPubKey: Uint8Array;
  verifyingKey: Uint8Array;
  depositTx: Transaction;
  vout: number;
};

const INITIAL_TIME_LOCK = 2000;

export class DepositService {
  private readonly config: WalletConfigService;
  private readonly connectionManager: ConnectionManager;

  constructor(
    config: WalletConfigService,
    connectionManager: ConnectionManager,
  ) {
    this.config = config;
    this.connectionManager = connectionManager;
  }

  private async validateDepositAddress({
    address,
    userPubkey,
  }: ValidateDepositAddressParams) {
    if (
      !address.depositAddressProof ||
      !address.depositAddressProof.proofOfPossessionSignature ||
      !address.depositAddressProof.addressSignatures
    ) {
      throw new Error(
        "proof of possession signature or address signatures is null",
      );
    }

    const operatorPubkey = subtractPublicKeys(address.verifyingKey, userPubkey);
    const msg = proofOfPossessionMessageHashForDepositAddress(
      await this.config.signer.getIdentityPublicKey(),
      operatorPubkey,
      address.address,
    );

    const taprootKey = p2tr(
      operatorPubkey.slice(1, 33),
      undefined,
      getNetwork(this.config.getNetwork()),
    ).tweakedPubkey;

    const isVerified = schnorr.verify(
      address.depositAddressProof.proofOfPossessionSignature,
      msg,
      taprootKey,
    );

    if (!isVerified) {
      throw new Error("proof of possession signature verification failed");
    }

    const addrHash = sha256(address.address);
    for (const operator of Object.values(this.config.getSigningOperators())) {
      if (operator.identifier === this.config.getCoordinatorIdentifier()) {
        continue;
      }

      const operatorPubkey = operator.identityPublicKey;
      const operatorSig =
        address.depositAddressProof.addressSignatures[operator.identifier];
      if (!operatorSig) {
        throw new Error("operator signature not found");
      }
      const sig = secp256k1.Signature.fromDER(operatorSig);

      const isVerified = secp256k1.verify(sig, addrHash, operatorPubkey);
      if (!isVerified) {
        throw new Error("signature verification failed");
      }
    }
  }

  async generateDepositAddress({
    signingPubkey,
  }: GenerateDepositAddressParams): Promise<GenerateDepositAddressResponse> {
    const sparkClient = await this.connectionManager.createSparkClient(
      this.config.getCoordinatorAddress(),
    );

    let depositResp: GenerateDepositAddressResponse;
    try {
      depositResp = await sparkClient.generate_deposit_address({
        signingPublicKey: signingPubkey,
        identityPublicKey: await this.config.signer.getIdentityPublicKey(),
        network: this.config.getNetworkProto(),
      });
    } catch (error) {
      throw new Error(`Error generating deposit address: ${error}`);
    }

    if (!depositResp.depositAddress) {
      throw new Error("No deposit address response from coordinator");
    }

    await this.validateDepositAddress({
      address: depositResp.depositAddress,
      userPubkey: signingPubkey,
    });

    return depositResp;
  }

  async createTreeRoot({
    signingPubKey,
    verifyingKey,
    depositTx,
    vout,
  }: CreateTreeRootParams) {
    // Create a root tx
    const rootTx = new Transaction();
    const output = depositTx.getOutput(vout);
    if (!output) {
      throw new Error("No output found in deposit tx");
    }
    const script = output.script;
    const amount = output.amount;
    if (!script || !amount) {
      throw new Error("No script or amount found in deposit tx");
    }

    rootTx.addInput({
      txid: getTxId(depositTx),
      index: vout,
    });

    rootTx.addOutput({
      script,
      amount,
    });

    const rootNonceCommitment =
      await this.config.signer.getRandomSigningCommitment();
    const rootTxSighash = getSigHashFromTx(rootTx, 0, output);

    // Create a refund tx
    const refundTx = new Transaction();
    const sequence = (1 << 30) | INITIAL_TIME_LOCK;
    refundTx.addInput({
      txid: getTxId(rootTx),
      index: 0,
      sequence,
    });

    const refundP2trAddress = getP2TRAddressFromPublicKey(
      signingPubKey,
      this.config.getNetwork(),
    );
    const refundAddress = btc
      .Address(getNetwork(this.config.getNetwork()))
      .decode(refundP2trAddress);
    const refundPkScript = btc.OutScript.encode(refundAddress);

    refundTx.addOutput({
      script: refundPkScript,
      amount: amount,
    });

    const refundNonceCommitment =
      await this.config.signer.getRandomSigningCommitment();
    const refundTxSighash = getSigHashFromTx(refundTx, 0, output);

    const sparkClient = await this.connectionManager.createSparkClient(
      this.config.getCoordinatorAddress(),
    );

    let treeResp: StartDepositTreeCreationResponse;

    try {
      treeResp = await sparkClient.start_deposit_tree_creation({
        identityPublicKey: await this.config.signer.getIdentityPublicKey(),
        onChainUtxo: {
          vout: vout,
          rawTx: depositTx.toBytes(),
          network: this.config.getNetworkProto(),
        },
        rootTxSigningJob: {
          rawTx: rootTx.toBytes(),
          signingPublicKey: signingPubKey,
          signingNonceCommitment: rootNonceCommitment,
        },
        refundTxSigningJob: {
          rawTx: refundTx.toBytes(),
          signingPublicKey: signingPubKey,
          signingNonceCommitment: refundNonceCommitment,
        },
      });
    } catch (error) {
      throw new Error(`Error starting tree creation: ${error}`);
    }

    if (!treeResp.rootNodeSignatureShares?.verifyingKey) {
      throw new Error("No verifying key found in tree response");
    }

    if (
      !treeResp.rootNodeSignatureShares.nodeTxSigningResult
        ?.signingNonceCommitments
    ) {
      throw new Error("No signing nonce commitments found in tree response");
    }

    if (
      !treeResp.rootNodeSignatureShares.refundTxSigningResult
        ?.signingNonceCommitments
    ) {
      throw new Error("No signing nonce commitments found in tree response");
    }

    if (
      !equalBytes(treeResp.rootNodeSignatureShares.verifyingKey, verifyingKey)
    ) {
      throw new Error("Verifying key does not match");
    }

    const rootSignature = await this.config.signer.signFrost({
      message: rootTxSighash,
      publicKey: signingPubKey,
      privateAsPubKey: signingPubKey,
      verifyingKey,
      selfCommitment: rootNonceCommitment,
      statechainCommitments:
        treeResp.rootNodeSignatureShares.nodeTxSigningResult
          .signingNonceCommitments,
      adaptorPubKey: new Uint8Array(),
    });

    const refundSignature = await this.config.signer.signFrost({
      message: refundTxSighash,
      publicKey: signingPubKey,
      privateAsPubKey: signingPubKey,
      verifyingKey,
      selfCommitment: refundNonceCommitment,
      statechainCommitments:
        treeResp.rootNodeSignatureShares.refundTxSigningResult
          .signingNonceCommitments,
      adaptorPubKey: new Uint8Array(),
    });

    const rootAggregate = await this.config.signer.aggregateFrost({
      message: rootTxSighash,
      statechainSignatures:
        treeResp.rootNodeSignatureShares.nodeTxSigningResult.signatureShares,
      statechainPublicKeys:
        treeResp.rootNodeSignatureShares.nodeTxSigningResult.publicKeys,
      verifyingKey: treeResp.rootNodeSignatureShares.verifyingKey,
      statechainCommitments:
        treeResp.rootNodeSignatureShares.nodeTxSigningResult
          .signingNonceCommitments,
      selfCommitment: createWasmSigningCommitment(rootNonceCommitment),
      publicKey: signingPubKey,
      selfSignature: rootSignature!,
      adaptorPubKey: new Uint8Array(),
    });

    const refundAggregate = await this.config.signer.aggregateFrost({
      message: refundTxSighash,
      statechainSignatures:
        treeResp.rootNodeSignatureShares.refundTxSigningResult.signatureShares,
      statechainPublicKeys:
        treeResp.rootNodeSignatureShares.refundTxSigningResult.publicKeys,
      verifyingKey: treeResp.rootNodeSignatureShares.verifyingKey,
      statechainCommitments:
        treeResp.rootNodeSignatureShares.refundTxSigningResult
          .signingNonceCommitments,
      selfCommitment: createWasmSigningCommitment(refundNonceCommitment),
      publicKey: signingPubKey,
      selfSignature: refundSignature,
      adaptorPubKey: new Uint8Array(),
    });

    let finalizeResp: FinalizeNodeSignaturesResponse;
    try {
      finalizeResp = await sparkClient.finalize_node_signatures({
        intent: SignatureIntent.CREATION,
        nodeSignatures: [
          {
            nodeId: treeResp.rootNodeSignatureShares.nodeId,
            nodeTxSignature: rootAggregate,
            refundTxSignature: refundAggregate,
          },
        ],
      });
    } catch (error) {
      throw new Error(`Error finalizing node signatures in deposit: ${error}`);
    }

    return finalizeResp;
  }
}
